src/libs/sys-util-rs/src/fs.rs

```diff
@@ -0,0 +1,165 @@
+    if !src_info.is_file() {
+        return Err(eother!("reflink_copy src {} is not a regular file", src));
+    }
+
+    // Make sure dst's parent exist. If dst is a regular file, then unlink it for later copy.
+    if dst_path.exists() {
+        let dst_info = fs::metadata(dst_path)?;
+        if !dst_info.is_file() {
+            return Err(eother!("reflink_copy dst {} is not a regular file", dst));
+        } else {
+            nix::unistd::unlink(dst_path)?;
+        }
+    } else if let Some(dst_parent) = dst_path.parent() {
+        if !dst_parent.exists() {
+            if let Err(e) = fs::create_dir_all(dst_parent) {
+                return Err(eother!(
+                    "reflink_copy: create_dir_all {} failed: {:?}",
+                    dst_parent.to_str().unwrap(),
+                    e
+                ));
+            }
+        } else {
+            let md = dst_parent.metadata()?;
+            if !md.is_dir() {
+                return Err(eother!("reflink_copy parent of {} is not a directory", dst));
+            }
+        }
+    }
+
+    // Reflink copy, and fallback to regular copy if reflink fails.
+    let src_file = fs::File::open(src_path)?;
+    let dst_file = fs::File::create(dst_path)?;
+    if let Err(e) = do_reflink_copy(src_file, dst_file) {
+        match e.raw_os_error() {
+            // Cross dev copy or filesystem doesn't support reflink, do regular copy
+            Some(os_err)
+                if os_err == nix::Error::EXDEV as i32
+                    || os_err == nix::Error::EOPNOTSUPP as i32 =>
+            {
+                warn!(
+                    sl!(),
+                    "reflink_copy: reflink is not supported ({:?}), do regular copy instead", e,
+                );
+                if let Err(e) = do_regular_copy(src.as_ref(), dst.as_ref()) {
+                    return Err(eother!(
+                        "reflink_copy: regular copy {} to {} failed: {:?}",
+                        src,
+                        dst,
+                        e
+                    ));
+                }
+            }
+            // Reflink copy failed
+            _ => {
+                return Err(eother!(
+                    "reflink_copy: copy {} to {} failed: {:?}",
+                    src,
+                    dst,
+                    e,
+                ))
+            }
+        }
+    }
+
+    Ok(())
+}
+
+// Copy file using cp command, which handles sparse file copy.
+fn do_regular_copy(src: &str, dst: &str) -> Result<()> {
+    match Command::new("cp")
+        .args(&["--sparse=auto", src, dst])
+        .output()
+    {
+        Ok(output) => match output.status.success() {
+            true => Ok(()),
+            false => Err(eother!("cp {} {} failed: {:?}", src, dst, output)),
+        },
+        Err(e) => Err(eother!("cp {} {} failed: {:?}", src, dst, e)),
+    }
+}
+
+/// Copy file by reflink
+fn do_reflink_copy(src: File, dst: File) -> Result<()> {
```

Comment:
We have some usage scenarios which cares about the time cost.
If it succeeds to do reflink copy, the copy operation itself should be fast enough, so we are trying to avoid an extra exec here.

Reply:
That sounds like an optimization that doesn't belong in the first set of commits.

Reply:
The whole reflink copy staff is included into the second round.